clang/lib/AST/ByteCode/Compiler.cpp

```diff
@@ -5312,6 +5314,147 @@ bool Compiler<Emitter>::VisitComplexUnaryOperator(const UnaryOperator *E) {
   return true;
 }
 
+template <class Emitter>
+bool Compiler<Emitter>::VisitVectorUnaryOperator(const UnaryOperator *E) {
+  const Expr *SubExpr = E->getSubExpr();
+  assert(SubExpr->getType()->isVectorType());
+
+  if (DiscardResult)
+    return this->discard(SubExpr);
+
+  auto prepareResult = [=]() -> bool {
+    if (!Initializing) {
+      std::optional<unsigned> LocalIndex = allocateLocal(SubExpr);
+      if (!LocalIndex)
+        return false;
+      return this->emitGetPtrLocal(*LocalIndex, E);
+    }
+    return true;
+  };
+
+  // The offset of the temporary, if we created one.
+  unsigned SubExprOffset = ~0u;
+  auto createTemp = [=, &SubExprOffset]() -> bool {
+    SubExprOffset = this->allocateLocalPrimitive(SubExpr, PT_Ptr, true, false);
+    if (!this->visit(SubExpr))
+      return false;
+    return this->emitSetLocal(PT_Ptr, SubExprOffset, E);
+  };
+
+  const auto *VecTy = SubExpr->getType()->getAs<VectorType>();
+  PrimType ElemT = classifyVectorElementType(SubExpr->getType());
+  auto getElem = [=](unsigned Offset, unsigned Index) -> bool {
+    if (!this->emitGetLocal(PT_Ptr, Offset, E))
+      return false;
+    return this->emitArrayElemPop(ElemT, Index, E);
+  };
+
+  switch (E->getOpcode()) {
+  case UO_Plus: // +x
+    return this->delegate(SubExpr);
+  case UO_Minus:
+    if (!prepareResult())
+      return false;
+    if (!createTemp())
+      return false;
+    for (unsigned I = 0; I != VecTy->getNumElements(); ++I) {
+      if (!getElem(SubExprOffset, I))
+        return false;
+      if (!this->emitNeg(ElemT, E))
+        return false;
+      if (!this->emitInitElem(ElemT, I, E))
+        return false;
+    }
+    break;
+  case UO_LNot: { // !x
+    if (!prepareResult())
+      return false;
+    if (!createTemp())
+      return false;
+
+    // In C++, the logic operators !, &&, || are available for vectors. !v is
+    // equivalent to v == 0.
+    // https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html
+    QualType SignedVecT =
+        Ctx.getASTContext().GetSignedVectorType(SubExpr->getType());
+    PrimType SignedElemT =
+        classifyPrim(SignedVecT->getAs<VectorType>()->getElementType());
```

Comment:
Instead of getting the types from the `ASTContext`, can't you get them from the type of `E`?

Reply:
It can be get from E, but it will produce a long and repeated if-else statements, so I want to reuse the original code.

Reply:
Why would that create a long if statement? Isn't the type we get from the `ASTContext` the same as the one we can extract from `E`?

I'm asking because it is always safer to use the result expression (`E` here), because that's what the surrounding code will use as well.

Reply:
Yeah, it's different, https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html says:
> In C++, the logic operators !, &&, || are available for vectors. !v is equivalent to v == 0, ....
and


> The result of the comparison is a vector of the same width and number of elements as the comparison operands with a signed integral element type.

Due to these, the type from `ASTContext` is different from the type from `E`. The type from `ASTContext` is a signed integer vector type, and the num of elements and the size of element type same as the vector type from `E`.

Reply:
I've just added
```c++
    assert(E->getType() == SignedVecT);
    assert(E->getType()->getAs<VectorType>()->getElementType() == SignedVecT->getAs<VectorType>()->getElementType());
```
and it doesn't break when compiling the `constexpr-vectors.cpp`. When are the two types different?

Reply:
Sorry, You are correct. I mistook E as a SubExpr of E. E's type should already has been set as a signed-vector type in Sema.